tests/test_end_to_end_zenodo.py

```diff
@@ -0,0 +1,196 @@
+# tests/test_end_to_end_zenodo.py
+import json
+import tempfile
+from pathlib import Path
+from copy import deepcopy
+from unittest import TestCase
+from unittest.mock import patch
+
+from django.core.management import call_command
+from django.test import override_settings
+
+from publications.models import Publication, Source
+
+
+class EndToEndZenodoTest(TestCase):
+    def setUp(self):
+        # Temp “project root”
+        self._tmpdir = tempfile.TemporaryDirectory()
+        self.project_root = Path(self._tmpdir.name)
+        self.templates_dir = self.project_root / "publications" / "templates"
+        self.cmds_dir = self.project_root / "publications" / "management" / "commands"
+        self.data_dir = self.project_root / "data"
+        self.templates_dir.mkdir(parents=True, exist_ok=True)
+        self.cmds_dir.mkdir(parents=True, exist_ok=True)
+        self.data_dir.mkdir(parents=True, exist_ok=True)
+
+        # Minimal README template with Sources
+        (self.templates_dir / "README.md.j2").write_text(
+            "# OPTIMAP FAIR Data Package\n"
+            "**Version:** {{ version }}\n\n"
+            "## Sources\n\n"
+            "{% for src in sources %}- [{{ src.name }}]({{ src.url }})\n{% endfor %}\n"
+            "\n## Codebook\n\n"
+            "| Field | Description |\n|---|---|\n| id | pk |\n",
+            encoding="utf-8",
+        )
+
+        # Test data
+        Publication.objects.create(title="A", publicationDate="2010-10-10")
+        # BAD label cases we want cleaned up (no DB NULLs!)
+        Source.objects.create(name="2000", url_field="http://optimap.science")    # numeric-only -> should become OPTIMAP
+        Source.objects.create(name="",     url_field="https://example.org")       # blank -> fallback to domain label
+        Source.objects.create(name=" ",    url_field="https://example.org")       # duplicate domain -> deduped
+        # Proper label kept
+        Source.objects.create(name="AGILE: GIScience Series", url_field="https://agile-giss.copernicus.org")
+
+        # Import command modules
+        import importlib
+        self.render_mod = importlib.import_module("publications.management.commands.render_zenodo_desc")
+        self.deploy_mod = importlib.import_module("publications.management.commands.deploy_zenodo")
+
+        # Fake Path so parents[3] stays inside tmp root
+        class FakePath(Path):
+            _flavour = Path(".")._flavour
+            def resolve(self):
+                return self
+        self.FakePath = FakePath
+        self.render_file = str(self.cmds_dir / "render_zenodo_desc.py")
+        self.deploy_file = str(self.cmds_dir / "deploy_zenodo.py")
+
+    def tearDown(self):
+        self._tmpdir.cleanup()
+
+    def test_end_to_end_render_then_deploy(self):
```

Comment:
This test case is too big. You should be able to split it up at least into 

1. file generation
2. deposition

You're welcome to copy and paste the file generation code from the first test to the second.